server/src/main/java/org/opensearch/gateway/remote/ClusterStateChecksum.java

```diff
@@ -0,0 +1,419 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.metadata.DiffableStringMap;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.common.io.stream.BufferedChecksumStreamOutput;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+import org.opensearch.core.xcontent.ToXContentFragment;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParseException;
+import org.opensearch.core.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import com.jcraft.jzlib.JZlib;
+
+import static org.opensearch.core.xcontent.XContentParserUtils.ensureExpectedToken;
+
+/**
+ * Stores checksum for all components in cluster state. This will be used to ensure cluster state is same across all nodes in the cluster.
+ */
+public class ClusterStateChecksum implements ToXContentFragment, Writeable {
+
+    private static final String ROUTING_TABLE_CS = "routing_table";
+    private static final String NODES_CS = "discovery_nodes";
+    private static final String BLOCKS_CS = "blocks";
+    private static final String CUSTOMS_CS = "customs";
+    private static final String COORDINATION_MD_CS = "coordination_md";
+    private static final String SETTINGS_MD_CS = "settings_md";
+    private static final String TRANSIENT_SETTINGS_MD_CS = "transient_settings_md";
+    private static final String TEMPLATES_MD_CS = "templated_md";
+    private static final String CUSTOM_MD_CS = "customs_md";
+    private static final String HASHES_MD_CS = "hashes_md";
+    private static final String INDICES_CS = "indices_md";
+    private static final String CLUSTER_STATE_CS = "cluster_state";
+    private static final Logger logger = LogManager.getLogger(ClusterStateChecksum.class);
+
+    long routingTableChecksum;
+    long nodesChecksum;
+    long blocksChecksum;
+    long clusterStateCustomsChecksum;
+    long coordinationMetadataChecksum;
+    long settingMetadataChecksum;
+    long transientSettingsMetadataChecksum;
+    long templatesMetadataChecksum;
+    long customMetadataMapChecksum;
+    long hashesOfConsistentSettingsChecksum;
+    long indicesChecksum;
+    long clusterStateChecksum;
+
+    public ClusterStateChecksum(ClusterState clusterState) {
```

Comment:
If we are anyways going with dedicated writeToSorted then should we write the checksums in their respective writeTo methods and call it `writeCheckSum` instead